mypy/subtypes.py

```diff
@@ -130,24 +143,33 @@ def visit_instance(self, left: Instance) -> bool:
         if isinstance(right, TupleType) and right.fallback.type.is_enum:
             return is_subtype(left, right.fallback)
         if isinstance(right, Instance):
-            # NOTO: left.type.mro may be None in quick mode if there
+            if (left, right) in right.type.cache:
+                return True
+            # NOTE: left.type.mro may be None in quick mode if there
             # was an error somewhere.
             if left.type.mro is not None:
                 for base in left.type.mro:
                     if base._promote and is_subtype(
                             base._promote, self.right, self.check_type_parameter,
                             ignore_pos_arg_names=self.ignore_pos_arg_names):
+                        right.type.cache.add((left, right))
```

Comment:
I am not super sure, but I think it is OK to always add them, since these are only for ``Instance``s and they (at least currently) are always used with ``ignore_pos_arg_names``.